### ISSUE TITLE
(Detection) I'm training custom data, having trouble with writing config.py. 

### ISSUE THREAD

**Issue by chungminho1** [OpenGVLab/InternImage#155]
I am trying to train my custom data, following https://github.com/OpenGVLab/InternImage/tree/master/detection.

```
├── data
    ├──train
    ├──test
    ├── annotations
        ├── train.coco.json
        ├── test.coco.json
├── config.py
├── configs
    ├── custom
        ├── config_instance.py
```

[config.py]

```
# dataset settings
dataset_type = 'CocoDataset'
data_root = 'data/'                                                                                                     #changed
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True, with_mask=True),
    dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels', 'gt_masks']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1333, 800),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    samples_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/train.coco.json',                                      #changed
        img_prefix=data_root + 'train/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/valid.coco.json',                                      #changed
        img_prefix=data_root + 'valid/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/valid.coco.json',                                      #changed
        img_prefix=data_root + 'valid/',
        pipeline=test_pipeline))
evaluation = dict(metric=['bbox', 'segm'], classwise=True)
```

[configs/custom/config_instance.py]

```
# --------------------------------------------------------
# InternImage
# Copyright (c) 2022 OpenGVLab
# Licensed under The MIT License [see LICENSE for details]
# --------------------------------------------------------
_base_ = [
    '../_base_/models/mask_rcnn_r50_fpn.py',
    '../../config.py',                                                                                                  #changed
    '../_base_/schedules/schedule_3x.py',
    '../_base_/default_runtime.py'
]
pretrained = 'https://huggingface.co/OpenGVLab/InternImage/resolve/main/internimage_s_1k_224.pth'
model = dict(
    backbone=dict(
        _delete_=True,
        type='InternImage',
        core_op='DCNv3',
        channels=80,
        depths=[4, 4, 21, 4],
        groups=[5, 10, 20, 40],
        mlp_ratio=4.,
        drop_path_rate=0.3,
        norm_layer='LN',
        layer_scale=1.0,
        offset_scale=1.0,
        post_norm=True,
        with_cp=False,
        out_indices=(0, 1, 2, 3),
        init_cfg=dict(type='Pretrained', checkpoint=pretrained)),
    neck=dict(
        type='FPN',
        in_channels=[80, 160, 320, 640],
        out_channels=256,
        num_outs=5))
# By default, models are trained on 8 GPUs with 2 images per GPU
data = dict(samples_per_gpu=2)
optimizer = dict(
    _delete_=True, type='AdamW', lr=0.0001, weight_decay=0.05,
    constructor='CustomLayerDecayOptimizerConstructor',
    paramwise_cfg=dict(num_layers=33, layer_decay_rate=1.0,
                       depths=[4, 4, 21, 4]))
optimizer_config = dict(grad_clip=None)
# fp16 = dict(loss_scale=dict(init_scale=512))
evaluation = dict(save_best='auto')
checkpoint_config = dict(
    interval=1,
    max_keep_ckpts=3,
    save_last=True,
)
```

And then I tried "bash dist_train.sh configs/custom/config_instance.py 2" ( I'm using 2 GPUs)
This is the result.
1 epoch is successfully trained but that's all.
What's the problem? I think I have to change more about my config files.
In dataset, I have only 1 class.
I really really need help and I would be really really thankful for your help!

```
/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launch                                                                                                             .py:186: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use_env is set by default in torchrun.
If your script expects `--local_rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See
https://pytorch.org/docs/stable/distributed.html#launch-utility for
further instructions

  FutureWarning,
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default                                                                                                             , to avoid your system being overloaded, please further tune the variable for op                                                                                                             timal performance in your application as needed.
*****************************************
/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/utils/setup_env.py                                                                                                             :49: UserWarning: Setting MKL_NUM_THREADS environment variable for each process                                                                                                              to be 1 in default, to avoid your system being overloaded, please further tune t                                                                                                             he variable for optimal performance in your application as needed.
  f'Setting MKL_NUM_THREADS environment variable for each process '
/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/utils/setup_env.py                                                                                                             :49: UserWarning: Setting MKL_NUM_THREADS environment variable for each process                                                                                                              to be 1 in default, to avoid your system being overloaded, please further tune t                                                                                                             he variable for optimal performance in your application as needed.
  f'Setting MKL_NUM_THREADS environment variable for each process '
2023-05-18 14:11:54,127 - mmdet - INFO - Environment info:
------------------------------------------------------------
sys.platform: linux
Python: 3.7.16 (default, Jan 17 2023, 22:20:44) [GCC 11.2.0]
CUDA available: True
GPU 0,1: NVIDIA GeForce RTX 4090
CUDA_HOME: None
GCC: gcc (Debian 10.2.1-6) 10.2.1 20210110
PyTorch: 1.11.0+cu113
PyTorch compiling details: PyTorch built with:
.
.
.
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
loading annotations into memory...
Done (t=0.00s)
creating index...
index created!
2023-05-18 14:11:58,834 - mmdet - INFO - Start running, host: root@0ce9050d4ee2, work_dir: /InternImage/detection/work_dirs/config_instance
2023-05-18 14:11:58,834 - mmdet - INFO - Hooks will be executed in the following order:
before_run:
(VERY_HIGH   ) StepLrUpdaterHook
(NORMAL      ) CheckpointHook
(LOW         ) DistEvalHook
(VERY_LOW    ) TextLoggerHook
 --------------------
before_train_epoch:
(VERY_HIGH   ) StepLrUpdaterHook
(NORMAL      ) NumClassCheckHook
(NORMAL      ) DistSamplerSeedHook
(LOW         ) IterTimerHook
(LOW         ) DistEvalHook
(VERY_LOW    ) TextLoggerHook
 --------------------
before_train_iter:
(VERY_HIGH   ) StepLrUpdaterHook
(LOW         ) IterTimerHook
(LOW         ) DistEvalHook
 --------------------
after_train_iter:
(ABOVE_NORMAL) OptimizerHook
(NORMAL      ) CheckpointHook
(LOW         ) IterTimerHook
(LOW         ) DistEvalHook
(VERY_LOW    ) TextLoggerHook
 --------------------
after_train_epoch:
(NORMAL      ) CheckpointHook
(LOW         ) DistEvalHook
(VERY_LOW    ) TextLoggerHook
 --------------------
before_val_epoch:
(NORMAL      ) NumClassCheckHook
(NORMAL      ) DistSamplerSeedHook
(LOW         ) IterTimerHook
(VERY_LOW    ) TextLoggerHook
 --------------------
before_val_iter:
(LOW         ) IterTimerHook
 --------------------
after_val_iter:
(LOW         ) IterTimerHook
 --------------------
after_val_epoch:
(VERY_LOW    ) TextLoggerHook
 --------------------
after_run:
(VERY_LOW    ) TextLoggerHook
 --------------------
2023-05-18 14:11:58,834 - mmdet - INFO - workflow: [('train', 1)], max: 12 epochs
2023-05-18 14:11:58,834 - mmdet - INFO - Checkpoints will be saved to /InternImage/detection/work_dirs/config_instance by HardDiskBackend.
2023-05-18 14:12:00,996 - mmdet - INFO - Saving checkpoint at 1 epochs
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 706/706, 83.6 task/s, elapsed: 8s, ETA:     0s

Traceback (most recent call last):
  File "./train.py", line 247, in <module>
    main()
  File "./train.py", line 243, in main
    meta=meta)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/apis/train.py", line 246, in train_detector
    runner.run(data_loaders, cfg.workflow)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 127, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 54, in train
    self.call_hook('after_train_epoch')
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 309, in call_hook
    getattr(hook, fn_name)(self)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmcv/runner/hooks/evaluation.py", line 267, in after_train_epoch
    self._do_evaluate(runner)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/core/evaluation/eval_hooks.py", line 135, in _do_evaluate
    key_score = self.evaluate(runner, results)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmcv/runner/hooks/evaluation.py", line 364, in evaluate
    results, logger=runner.logger, **self.eval_kwargs)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/datasets/coco.py", line 641, in evaluate
    result_files, tmp_dir = self.format_results(results, jsonfile_prefix)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/datasets/coco.py", line 383, in format_results
    result_files = self.results2json(results, jsonfile_prefix)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/datasets/coco.py", line 320, in results2json
    json_results = self._segm2json(results)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/mmdet/datasets/coco.py", line 271, in _segm2json
    data['category_id'] = self.cat_ids[label]
IndexError: list index out of range
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 3342 closing signal SIGTERM
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 0 (pid: 3341) of binary: /opt/conda/envs/internimage/bin/python
Traceback (most recent call last):
  File "/opt/conda/envs/internimage/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/envs/internimage/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launch.py", line 193, in <module>
    main()
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launch.py", line 189, in main
    launch(args)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launch.py", line 174, in launch
    run(args)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/run.py", line 718, in run
    )(*cmd_args)
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launcher/api.py", line 131, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/conda/envs/internimage/lib/python3.7/site-packages/torch/distributed/launcher/api.py", line 247, in launch_agent
    failures=result.failures,
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
./train.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2023-05-18_14:12:12
  host      : 0ce9050d4ee2
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 3341)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

**Comment by KevinXu-01** [comment]
Hi, if you want to train on your custom data, you might want to change the CLASSES settings in mmdet/datasets/coco.py and mmdet/core/evaluation/class_names.py
The mmdet folder can be found in your virtual python env.

**Comment by zrr1005** [comment]
I also encounteredI also encountered the same problem when training ADE20KDataSet. How did you finally solve it the same problem when training ADE20KDataSet. How did you finally solve it

**Comment by Nguyen-duc-quyen** [comment]
> I also encounteredI also encountered the same problem when training ADE20KDataSet. How did you finally solve it the same problem when training ADE20KDataSet. How did you finally solve it

Hi, I also encountered a similar problem when I trained on my custom dataset. After spending some time debugging, the culprit of my problem was that I named the categories wrong. The original classes' names were ['Car', 'Bus', .. etc] but I changed the CLASSES in the mmdet/datasets/coco.py as follows: ['car', 'bus', ... etc]. As a result, there was a mismatch, causing the getCatIds () method to return [ ] since it could not find any of the classes. I hope this helps.
